File: scripts/package.ts
#!/usr/bin/env bun
import { $ } from "bun";
import { mkdtemp } from "node:fs/promises";
import { tmpdir } from "node:os";
import { basename, join, resolve } from "node:path";

async function writeChecksum(file: string, output: string) {
  const hash = Bun.SHA256.hash(await Bun.file(file).bytes(), "hex");
  await Bun.write(output, `${hash}  ${basename(file)}\n`);
}

/** Archive a published executable with documentation, licenses, and checksums. */
export async function packageExecutable(rid: string, root = resolve(import.meta.dir, "..")) {
  if (!/^[a-z0-9]+(?:-[a-z0-9]+)+$/.test(rid)) {
    throw new Error(`Invalid runtime identifier: ${rid}`);
  }
  const windows = rid.startsWith("win-");
  const binaryName = windows ? "yt-dlp-explode.exe" : "yt-dlp-explode";
  const binary = join(root, "artifacts", rid, binaryName);
  const dist = join(root, "artifacts", "dist");
  const name = `yt-dlp-explode-${rid}`;
  const archive = join(dist, `${name}.${windows ? "zip" : "tar.gz"}`);
  const temporary = await mkdtemp(join(tmpdir(), "explode-package-"));
  try {
    const directory = join(temporary, name);
    await $`mkdir -p ${dist} ${join(directory, "licenses")}`;
    await $`cp -p ${binary} README.md COMPATIBILITY.md LICENSE ${directory}`.cwd(root);
    await $`cp -p external/YoutubeExplode/License.txt ${join(directory, "licenses", "YoutubeExplode.txt")}`.cwd(root);
    await writeChecksum(binary, join(directory, "SHA256SUMS"));
    await $`tar ${windows ? "-acf" : "-czf"} ${archive} -C ${temporary} ${name}`;
  } finally {
    await $`rm -rf ${temporary}`;
  }
  await writeChecksum(archive, `${archive}.sha256`);
  return archive;
}

if (import.meta.main) {
  if (Bun.argv.length !== 3 || ["-h", "--help"].includes(Bun.argv[2])) {
    console.log("Usage: bun scripts/package.ts <runtime-identifier>");
    process.exit(Bun.argv.length === 3 ? 0 : 2);
  }
  console.log(await packageExecutable(Bun.argv[2]));
}

File: scripts/package.py
#!/usr/bin/env python3
"""Archive a published executable with documentation, licenses, and a checksum."""

import hashlib
from pathlib import Path
import shutil
import sys
import tarfile
import tempfile
import zipfile


root = Path(__file__).resolve().parent.parent
rid = sys.argv[1]
binary_name = "yt-dlp-explode.exe" if rid.startswith("win-") else "yt-dlp-explode"
binary = root / "artifacts" / rid / binary_name
if not binary.is_file():
    raise SystemExit(f"Missing published executable: {binary}")

dist = root / "artifacts/dist"
dist.mkdir(parents=True, exist_ok=True)
name = f"yt-dlp-explode-{rid}"
with tempfile.TemporaryDirectory() as temporary:
    package = Path(temporary) / name
    package.mkdir()
    shutil.copy2(binary, package / binary_name)
    for filename in ("README.md", "LICENSE"):
        shutil.copy2(root / filename, package / filename)
    shutil.copytree(root / "licenses", package / "licenses")
    (package / "SHA256SUMS").write_text(
        f"{hashlib.sha256(binary.read_bytes()).hexdigest()}  {binary_name}\n",
        encoding="utf-8",
    )
    if rid.startswith("win-"):
        archive = dist / f"{name}.zip"
        with zipfile.ZipFile(archive, "w", zipfile.ZIP_DEFLATED) as output:
            for path in sorted(package.rglob("*")):
                if path.is_file():
                    output.write(path, path.relative_to(package.parent))
    else:
        archive = dist / f"{name}.tar.gz"
        with tarfile.open(archive, "w:gz") as output:
            output.add(package, arcname=name)

(dist / f"{archive.name}.sha256").write_text(
    f"{hashlib.sha256(archive.read_bytes()).hexdigest()}  {archive.name}\n", encoding="utf-8"
)
print(archive)
